File: 647-palindrome/js/647.js
#!/usr/bin/env node

// 60ms, 35MB on Leetcode - much better than 647-dynamic.js
// runtime 94th percentile, memory 50th percentile
const expand_palin_count = (s, i, j) => {
    let count = 0;
    while (i >= 0 && j < s.length && s[i] == s[j]) {
        count += 1;
        i -= 1;
        j += 1;
    }
    return count;
};
const countSubstrings = s => {
    let count = 0;
    for (let i = 0; i < s.length; i++) {
        count += expand_palin_count(s, i, i);
        count += expand_palin_count(s, i, i+1);
    }
    return count;
};

let actualResult = countSubstrings("aaa");
let expectedResult = 6;
console.log("Expected result:", expectedResult);
console.log("Actual result:", actualResult);
let match = JSON.stringify(actualResult) == JSON.stringify(expectedResult);
console.log("Pass?", match);

File: 647-palindrome/elixir/647.exs
#! /usr/bin/env elixir
defmodule Solution do
  def expand_palin_count(s, i, j) do
    cond do
      i >= 0 && j < String.length(s) && String.at(s, i) == String.at(s, j) ->
        1 + expand_palin_count(s, i - 1, j + 1)

      true ->
        0
    end
  end

  def count_substrings(s) do
    0..String.length(s)
    |> Enum.map(fn i -> expand_palin_count(s, i, i) + expand_palin_count(s, i, i + 1) end)
    |> Enum.sum()
  end
end

IO.puts("Expect to see: 6")

Solution.count_substrings("aaa")
|> IO.inspect()
